hudi-client/hudi-client-common/src/main/java/org/apache/hudi/index/simple/HoodieGlobalSimpleIndex.java

```diff
@@ -68,21 +68,19 @@ public <R> HoodieData<HoodieRecord<R>> tagLocation(
   protected <R> HoodieData<HoodieRecord<R>> tagLocationInternal(
       HoodieData<HoodieRecord<R>> inputRecords, HoodieEngineContext context,
       HoodieTable hoodieTable) {
-    List<Pair<String, HoodieBaseFile>> latestBaseFiles = getAllBaseFilesInTable(context, hoodieTable);
+    HoodieData<Pair<String, HoodieBaseFile>> latestBaseFiles = getAllBaseFilesInTable(context, hoodieTable);
     HoodiePairData<String, HoodieRecordGlobalLocation> allKeysAndLocations =
-        fetchRecordGlobalLocations(context, hoodieTable, config.getGlobalSimpleIndexParallelism(), latestBaseFiles);
+        fetchRecordGlobalLocations(hoodieTable, config.getGlobalSimpleIndexParallelism(), latestBaseFiles);
     boolean mayContainDuplicateLookup = hoodieTable.getMetaClient().getTableType() == MERGE_ON_READ;
     boolean shouldUpdatePartitionPath = config.getGlobalSimpleIndexUpdatePartitionPath() && hoodieTable.isPartitioned();
     return tagGlobalLocationBackToRecords(inputRecords, allKeysAndLocations,
         mayContainDuplicateLookup, shouldUpdatePartitionPath, config, hoodieTable);
   }
 
   private HoodiePairData<String, HoodieRecordGlobalLocation> fetchRecordGlobalLocations(
-      HoodieEngineContext context, HoodieTable hoodieTable, int parallelism,
-      List<Pair<String, HoodieBaseFile>> baseFiles) {
-    int fetchParallelism = Math.max(1, Math.min(baseFiles.size(), parallelism));
-
-    return context.parallelize(baseFiles, fetchParallelism)
+      HoodieTable hoodieTable, int parallelism,
+      HoodieData<Pair<String, HoodieBaseFile>> baseFiles) {
+    return baseFiles.repartition(Math.max(1, Math.min(baseFiles.getNumPartitions(), parallelism)))
```

Comment:
Getting the size would require you to do a count and collect a result on the driver. A lot of the changes here were to avoid collecting intermediate results on the driver.

Reply:
I guess we may need to do that though to maintain parity. Right now the number of incoming partitions for the dataset is equivalent to the number partitions in the Hudi table. This code would previously split each file to its own partition essentially

Reply:
baseFiles.getNumPartitions() likely is equivalent to number of physical hudi partitions. 
with partitions having large no of file groups, this might have perf impact.